libs/auth/react/src/lib/Authenticator/CheckIdpSession.tsx

```diff
@@ -30,58 +22,108 @@ export interface SessionInfoResponse {
   isExpired?: boolean
 }
 
-interface SessionInfoRequest {
-  type: SessionInfoMessageType.SessionInfoRequest
+interface UserSession {
+  /* Flag to indicate if we have received session info message from the IDS iframe. */
+  hasMessage: boolean
+
+  /* The expected time when the user session is ending. */
+  sessionEnd: Date | null
+
+  /* The handle of the setInterval function. If set this indicates the user session active state. */
+  intervalHandle: ReturnType<typeof setInterval> | null
+
+  /* Number of times we have tried to load the iframe to receive a new session info message. */
+  retryCount: number
 }
 
-const messageEventName = 'message'
+const MAX_RETRIES = 2
+const MESSAGE_EVENT_NAME = 'message'
+const ACTIVE_SESSION_DELAY = 5 * 1000
+const CHECK_SESSION_INTERVAL = 2 * 1000
+
+const EMPTY_SESSION: UserSession = {
+  retryCount: 0,
+  sessionEnd: null,
+  hasMessage: false,
+  intervalHandle: null,
+}
 
 export const CheckIdpSession = () => {
   const userManager = getUserManager()
   const authSettings = getAuthSettings()
   const iframeSrc = `${authSettings.authority}${authSettings.checkSessionPath}`
-  const [iframeChecksum, setIframeChecksum] = useState(0)
-  const sessionTimeout = useRef<ReturnType<typeof setTimeout> | null>(null)
+  const [iframeId, reloadIframe] = useReducer((id) => id + 1, 0)
+  const userSession = useRef<UserSession>({ ...EMPTY_SESSION })
+
+  const isActive = useCallback(() => {
+    return !!userSession.current.intervalHandle
+  }, [])
+
+  const resetUserSession = useCallback(() => {
+    if (userSession.current.intervalHandle) {
+      clearInterval(userSession.current.intervalHandle)
+    }
+    userSession.current.intervalHandle = null
+    userSession.current.hasMessage = false
+    userSession.current.retryCount = 0
+    // Intentionally not resetting sessionEnd as it
+    // indicates that the user has had session before.
+  }, [])
 
   const signInRedirect = useCallback(async () => {
     await userManager.removeUser()
     return window.location.reload()
   }, [userManager])
 
-  const onLoadHandler = (event: SyntheticEvent<HTMLIFrameElement>) => {
-    if (authSettings.authority) {
-      event?.currentTarget?.contentWindow?.postMessage(
-        {
-          type: SessionInfoMessageType.SessionInfoRequest,
-        } as SessionInfoRequest,
-        authSettings.authority,
-      )
-    }
-  }
+  const checkActiveSession = useCallback(() => {
+    setTimeout(() => {
+      const hasBeenActive = !!userSession.current.sessionEnd
+      const { retryCount, hasMessage } = userSession.current
+
+      if (!isActive() && retryCount > MAX_RETRIES && hasBeenActive) {
+        window.location.reload()
+      } else if (!hasMessage && retryCount < MAX_RETRIES) {
```

Comment:
Would be nice to reuse the `isActive()` and get rid of `hasMessage`, but I understand if you were worried about a race condition with `sessionInfo.isExpired`.

If that's a concern you should also toggle `hasMessage` if there is "No user session".

Reply:
When can a session info message be active but not have an `expiresIn` value?